docs/reactive-nonblocking.qmd

```diff
@@ -0,0 +1,316 @@
+Because CPU-bound or synchronous tasks will block the main thread, we're back to where we started: the rest of the application cannot proceed until the task completes.
+
+Fortunately, we can rely on Python's built-in support for running asyncio tasks on different threads or processes.
+
+::: {.callout-note}
+The examples below work well in Shiny Core, but a naive port to Shiny Express will not work as well.
+The `ThreadPoolExecutor` and `ProcessPoolExecutor` objects need to be created as module-level variables, not as session-level variables, since we'd ideally like to pool resources across all sessions.
+:::
+
+::: {.callout-note}
+`ProcessPoolExecutor` is not available in Shinylive (e.g. WASM mode).
+`ThreadPoolExecutor` is available in Shinylive and appears to work, but doesn't: it actually performs all of its work on the main thread.
+:::
+
+The following example shows how to run a task on a different thread.
+This is a good strategy for code that does synchronous I/O, like reading from disk, a database, or a remote API endpoint.
+It's not as good of a strategy for CPU-bound code, because Python's [global interpreter lock](https://realpython.com/python-gil/) will prevent the task from running concurrently with other Python code.
+
+```{python}
+#| eval: false
+import asyncio
+import concurrent.futures
+import time
+
+from shiny import App, reactive, render, ui
+
+app_ui = ui.page_fluid(
+    ui.input_numeric("x", "x", value=1),
+    ui.input_numeric("y", "y", value=2),
+    ui.input_task_button("btn", "Add numbers"),
+    ui.output_text("sum"),
+)
+
+pool = concurrent.futures.ThreadPoolExecutor()
+
+def slow_sum(x, y):
+    time.sleep(5)  # Simulate a slow synchronous task
+    return x + y
+
+def server(input, output, session):
+    @ui.bind_task_button(button_id="btn")
+    @reactive.extended_task
+    async def sum_values(x, y):
+        loop = asyncio.get_event_loop()
+        return await loop.run_in_executor(pool, slow_sum, x, y)
+
+    @reactive.effect
+    @reactive.event(input.btn)
+    def btn_click():
+        sum_values(input.x(), input.y())
+
+    @render.text
+    def sum():
+        return str(sum_values.result())
+
+app = App(app_ui, server)
+app.on_shutdown(pool.shutdown)
+```
+
+With a small tweak, we can run the task on a different process instead of a different thread.
+This is a good strategy for CPU-bound code, because it allows the task to run concurrently with other Python code.
+
+```{python}
+#| eval: false
+import asyncio
+import concurrent.futures
+import time
+
+from shiny import App, reactive, render, ui
+
+app_ui = ui.page_fluid(
+    ui.input_numeric("x", "x", value=1),
+    ui.input_numeric("y", "y", value=2),
+    ui.input_task_button("btn", "Add numbers"),
+    ui.output_text("sum"),
+)
+
+pool = concurrent.futures.ProcessPoolExecutor()
```

Comment:
Feels like it's be better UX to just callout this change rather than listing the entire example again? Or put both examples into a tabset?

Reply:
Good call, especially since they don't actually run in the page.